File: webapp/health-monitor/mobile/src/api/endpoints.ts
import { http } from "./client";
import type {
  Activity,
  BloodPressure,
  CorrelationResult,
  DailyMetric,
  Summary,
  Survey,
} from "./types";

const { request, qs } = http;

export const api = {
  getSummary: () => request<Summary>("/summary"),

  getMetrics: (params: { source?: string; days?: number } = {}) =>
    request<DailyMetric[]>(`/metrics${qs(params)}`),
  postMetrics: (body: DailyMetric) =>
    request<{ id: number }>("/metrics", {
      method: "POST",
      body: JSON.stringify(body),
    }),

  getSurveys: (limit = 30) => request<Survey[]>(`/surveys${qs({ limit })}`),
  postSurvey: (body: Partial<Survey>) =>
    request<{ id: number }>("/surveys", {
      method: "POST",
      body: JSON.stringify(body),
    }),
  deleteSurvey: (id: number) =>
    request<null>(`/surveys/${id}`, { method: "DELETE" }),

  getBloodPressure: (days = 30) =>
    request<BloodPressure[]>(`/blood-pressure${qs({ days })}`),
  postBloodPressure: (body: Partial<BloodPressure>) =>
    request<{ id: number }>("/blood-pressure", {
      method: "POST",
      body: JSON.stringify(body),
    }),

  getActivities: (params: { sport_type?: string; days?: number; limit?: number } = {}) =>
    request<Activity[]>(`/activities${qs(params)}`),
  getSportTypes: () => request<string[]>("/activities/sport-types"),
  getActivitySummary: (params: { days?: number } = {}) =>
    request<{ by_sport: unknown[]; weekly: unknown[] }>(`/activities/summary${qs(params)}`),

  getCorrelation: (params: { days?: number } = {}) =>
    request<CorrelationResult>(`/analysis/correlation${qs(params)}`),

  health: () => request<{ status: string }>("/health"),
};

File: webapp/health-monitor/mobile/lib/api/endpoints.dart
import 'client.dart';
import 'types.dart';

class Api {
  Api._();

  static Future<Summary> getSummary() =>
      apiRequest('/summary', parse: (j) => Summary.fromJson(j as Map<String, dynamic>));

  static Future<List<DailyMetric>> getMetrics({String? source, int? days}) => apiRequest(
        '/metrics${qs({'source': source, 'days': days})}',
        parse: (j) => (j as List)
            .map((e) => DailyMetric.fromJson(e as Map<String, dynamic>))
            .toList(),
      );

  static Future<List<Survey>> getSurveys({int limit = 30}) => apiRequest(
        '/surveys${qs({'limit': limit})}',
        parse: (j) =>
            (j as List).map((e) => Survey.fromJson(e as Map<String, dynamic>)).toList(),
      );

  static Future<void> postSurvey(Map<String, dynamic> body) => apiRequest(
        '/surveys',
        method: 'POST',
        body: body,
        parse: (_) {},
      );

  static Future<void> deleteSurvey(int id) => apiRequest(
        '/surveys/$id',
        method: 'DELETE',
        parse: (_) {},
      );

  static Future<List<Activity>> getActivities({String? sportType, int? days, int? limit}) =>
      apiRequest(
        '/activities${qs({'sport_type': sportType, 'days': days, 'limit': limit})}',
        parse: (j) =>
            (j as List).map((e) => Activity.fromJson(e as Map<String, dynamic>)).toList(),
      );

  static Future<CorrelationResult> getCorrelation({int? days}) => apiRequest(
        '/analysis/correlation${qs({'days': days})}',
        parse: (j) => CorrelationResult.fromJson(j as Map<String, dynamic>),
      );

  static Future<Map<String, dynamic>> health() => apiRequest(
        '/health',
        parse: (j) => j as Map<String, dynamic>,
      );
}
